### ISSUE TITLE
block-plus-minus triggers warnings about overwriting block definitions

### ISSUE THREAD

**Issue by moniika** [RaspberryPiFoundation/blockly-samples#768]
**Category**

Plugins

**Component**
block-plus-minus

**Describe the bug**
Warning in console about overwriting existing json definitions.
```
Block definition #0 in JSON array overwrites prior definition of "lists_create_with".
Block definition #0 in JSON array overwrites prior definition of "procedures_defnoreturn".
Block definition #0 in JSON array overwrites prior definition of "procedures_defreturn".
```

**To Reproduce**

1. Run plugin using `npm run start`
2. Observe warnings in console.

**Expected behavior**

No warnings in console.

**Additional context**

This could be fixed by adding APIs in core to allow for overriding exising block definitions.

**Comment by BeksOmega** [comment]
Related to https://github.com/google/blockly/issues/2788

**Comment by johnnyoshika** [comment]
@BeksOmega Is this happening because keys are not unregistering before overriding the default block behaviors? I noticed that for `controls_if_mutator` and `text_join_mutator`, their keys gets unregistered before registering, so a warning is not logged to the console for those:

https://github.com/google/blockly-samples/blob/master/plugins/block-plus-minus/src/if.js#L178-L180
https://github.com/google/blockly-samples/blob/master/plugins/block-plus-minus/src/text_join.js

`lists_create_with`, `procedures_defnoreturn`, and `procedures_defreturn` on the other hand do not unregister their keys.

**Comment by BeksOmega** [comment]
> @BeksOmega Is this happening because keys are not unregistering before overriding the default block behaviors? I noticed that for `controls_if_mutator` and `text_join_mutator`, their keys gets unregistered before registering, so a warning is not logged to the console for those:
> 
> https://github.com/google/blockly-samples/blob/master/plugins/block-plus-minus/src/if.js#L178-L180
> https://github.com/google/blockly-samples/blob/master/plugins/block-plus-minus/src/text_join.js
> 
> `lists_create_with`, `procedures_defnoreturn`, and `procedures_defreturn` on the other hand do not unregister their keys.

Yeah definitely! So mutators/extensions have the ability to be unregistered, but blocks do not (currently) because they're applied directly to the `Blockly.Blocks` object, rather than being provided via the registry.

The past decision (noted [here](https://github.com/google/blockly/issues/2788#issuecomment-534278001)) was to leave overriding blocks as a warning, and to not provide the ability to unregister them. But with plugins that decision may need to be reevaluated. If you have thoughts on it feel free to leave them here :D

**Comment by johnnyoshika** [comment]
Thanks @BeksOmega for the explanation. I agree, it would be nice if we can get rid of these warnings some day. For example in my use case, I run unit tests in my continuous deployment process and test my custom blocks and logic. I see dozens of these warnings on every run, which can get frustrating as it makes it hard to see other errors and warnings (if there are any). Here's a partial screenshot of the continuous deployment output:

![image](https://user-images.githubusercontent.com/504505/126598850-3d7c1153-3e6e-42d4-ab56-c4c38f47ab93.png)

**Comment by BeksOmega** [comment]
@johnnyoshika 

I think we might be able to fix this by explicitly deleting the blocks before defining them with json. Eg:
```
// Add this line.
delete Blockly.Blocks['procedures_defnoreturn'];

Blockly.defineBlocksWithJsonArray([
  {
    "type": "procedures_defnoreturn",
    // etc...
  }
]);
```

But I'm not sure. If anyone wants to give it a shot and put up a PR we'd really appreciate it! Otherwise I have a note in my task manager to try it when I have time =)

Thanks for posting more info, including your console logs, @johnnyoshika Those were really helpful for seeing the scope of the problem!

**Comment by johnnyoshika** [comment]
@BeksOmega That did the trick! The warnings went away once I deleted the default blocks.

I wasn't sure how best to test it, so these are the steps I went through after cloning this repo:

```
npm install
npm run build
npm run boot
cd plugins/block-plus-minus
npm start
```

At this point, I was able to test the +/- blocks at http://localhost:3000/test/ and I saw the 3 warnings specific to this issue there (among many other warnings).

I added these in `procedures.js`:

```
delete Blockly.Blocks['procedures_defreturn'];
delete Blockly.Blocks['procedures_defnoreturn'];
```

...and this in `list_create.js`

```
delete Blockly.Blocks['lists_create_with'];
```

When I loaded http://localhost:3000/test/ after these changes the 3 warnings specific to this issue went away.

To ensure the changes worked properly in a real-world app, I rebuilt the project (`npm run build` at the root of the repo) to re-generate the contents of the `plugins/block-plus-minus/dist` folder. I then copied the contents of `plugins/block-plus-minus/dist` to my own app's `my-app/node_modules/@blockly/block-plus-minus/dist` and tested my app. The warnings went away and everything continued to work as expected.

If there's a better way to test these plugins, please let me know.

I'll create a PR shortly.

**Comment by johnnyoshika** [comment]
Opened PR here: https://github.com/google/blockly-samples/pull/829

**Comment by BeksOmega** [comment]
Closed by #829
